tests/e2e/test_parameterized_queries.py

```diff
@@ -0,0 +1,382 @@
+            with conn.cursor() as cursor:
+                to_return = cursor.execute(SELECT_QUERY).fetchone()
+            with conn.cursor() as cursor:
+                cursor.execute(DELETE_QUERY)
+
+        return to_return
+
+    def _native_roundtrip(
+        self, parameters: Union[Dict, List[Dict]], bypass_patch=False
+    ):
+        with self.connection(extra_params={"use_inline_params": False}) as conn:
+            with conn.cursor() as cursor:
+                with self.conditional_protocol_patch(bypass_patch):
+                    cursor.execute(self.NATIVE_QUERY, parameters=parameters)
+                    return cursor.fetchone()
+
+    def _get_one_result(self, approach: ParameterApproach, params, bypass_patch=False):
+        """When approach is INLINE then we use %(param)s paramstyle and a connection with use_inline_params=True
+        When approach is NATIVE then we use :param paramstyle and a connection with use_inline_params=False
+        """
+
+        if approach == ParameterApproach.INLINE:
+            return self._inline_roundtrip(params)
+        elif approach == ParameterApproach.NATIVE:
+            return self._native_roundtrip(params, bypass_patch)
+
+    def _quantize(self, input: Union[float, int], place_value=2) -> Decimal:
+        return Decimal(str(input)).quantize(Decimal("0." + "0" * place_value))
+
+    def test_compute_says_it_doesnt_support_native_params(self):
+        """This is a canary test case that will should be removed when it begins to fail.
+
+        It asserts that the target compute resource returns a protocol_version that _should_ mean it can't support
+        ParameterApproach.NATIVE.
+
+        We added this test case because the DBR we use for testing returns a protocol version that is too
+        low to support the native approach, but it does in fact support this approach. Without this patch
+        we'd have no way to run tests of the native approach because the connector would raise an exception.
+
+        Once the target compute accurately advertises the right protocol version, this test will begin
+        to fail and we can remove the conditional_protocol_patch.
+        """
+        assert (
+            self.compute_says_it_doesnt_support_native_params()
+        ), "Compute no longer says it doesn't support native params. Remove conditional_protocol_patch from this file."
+
+    @patch(
+        "databricks.sql.client.Connection.server_parameterized_queries_enabled",
+        return_value=False,
+    )
+    def test_protocol_too_low(self, mock_parameterized_queries_enabled):
+        params = {"p": None}
+        with pytest.raises(
+            NotSupportedError,
+            match="Parameterized operations are not supported by this server. DBR 14.1 is required.",
+        ):
+            result = self._get_one_result(
+                ParameterApproach.NATIVE, params, bypass_patch=True
+            )
+
+    @both_approaches
```

Comment:
Any test that is not decorated with `@both_approaches` is effectively unchanged in this PR, apart from updating the call to `_get_one_result` to include a `ParameterApproach`.